rpc/wrtc_base_channel.go

```diff
@@ -61,101 +61,99 @@ func newBaseChannel(
-				// its internal resources. Notably, things like `TrackRemote.ReadRTP` can hang. We'd
-				// instead prefer for that to receive an EOF, so it can close normally. Therefore
-				// upon reaching the `Disconnected` state, we explicitly call `Close` on our side of
-				// the `PeerConnection`.
-				//
-				// We chose here to call close for all cases of `Disconnected`, `Failed` and
-				// `Closed`. We rely on pion's `PeerConnection.GracefulClose` method being idempotent.
-				// We "gracefully" close to wait for the read loop to complete.
-				if err := peerConn.GracefulClose(); err != nil {
-					logger.Debugw("Error closing peer connection",
-						"conn_id", currConnID,
-						"conn_state", connectionState.String(),
-						"err", err,
-					)
+			}
+
+			// We will close+wait on all of the channel related resources. We will additionally
+			// close the PeerConnection, but not wait on that to drain its resources. Any desired
+			// PeerConnection waiting will be performed by the client connection object or server
+			// object.
+			ch.Close()
+			peerDoneOnce.Do(func() {
+				if onPeerDone != nil {
+					onPeerDone()
 				}
-			case webrtc.ICEConnectionStateConnected:
-				if onICEConnected != nil {
+			})
+		case webrtc.ICEConnectionStateConnected:
+			if onICEConnected != nil {
+				// The user of `onICEConnected` waits for a few seconds before enumerating/logging
+				// candidates. Spin off a goroutine to avoid consequences of sleeping.
+				ch.activeBackgroundWorkers.Add(1)
+				go func() {
+					defer ch.activeBackgroundWorkers.Done()
 					onICEConnected()
-				}
-				fallthrough
-			case webrtc.ICEConnectionStateChecking, webrtc.ICEConnectionStateCompleted,
-				webrtc.ICEConnectionStateNew:
-				fallthrough
-			default:
-				candPair, hasCandPair := webrtcPeerConnCandPair(peerConn)
-				connInfo := getWebRTCPeerConnectionStats(peerConn)
-				connIDMu.Lock()
-				connID = connInfo.ID
-				connIDMu.Unlock()
-				connectionStateChangedLogFields := []interface{}{
-					"conn_id", connInfo.ID,
-					"conn_local_candidates", connInfo.LocalCandidates,
-					"conn_remote_candidates", connInfo.RemoteCandidates,
-				}
-				if hasCandPair {
-					// Use info level when there is a selected candidate pair, as a
-					// connection has been established.
-					connectionStateChangedLogFields = append(connectionStateChangedLogFields,
-						"candidate_pair", candPair.String())
-					logger.Infow("Connection establishment succeeded", connectionStateChangedLogFields...)
-				} else {
-					// Use debug level when there is no selected candidate pair to avoid
-					// noise.
-					connectionStateChangedLogFields = append(connectionStateChangedLogFields,
-						"conn_state", connectionState.String())
-					logger.Debugw("Connection state changed", connectionStateChangedLogFields...)
-				}
+				}()
 			}
-		})
-	}
-	peerConn.OnICEConnectionStateChange(connStateChanged)
-
-	// fire once
-	connStateChanged(peerConn.ICEConnectionState())
```

Comment:
I claim that for the cases where this is a observable behavior change, those are the cases we're uninterested in.

Reply:
> // fire once

Yeah... I don't even know what this was doing.